docs/rfc/rfc-002-zero-copy-encoding.md

```diff
@@ -0,0 +1,402 @@
+   field detector. Furthermore, in [ADR 054](../architecture/adr-054-semver-compatible-modules.md) this "feature" of protobuf
+   poses one of the biggest problems for correct forward compatibility between different versions of the same module.
+2) not allowing new fields in existing messages makes the generated code in languages like Rust (which is currently our
+   highest priority target), much simpler and more performant because we can assume a fixed size struct gets allocated.
+   If new fields can be added to existing messages, we need to encode the number of fields into the message and then
+   do runtime checks. So this both increases memory layers and requires another layout of indirection. With the encoding
+   proposed below, "plain old Rust structs" (used with some special field types) can be used.
+
+Instead of adding new fields to existing messages, APIs can add new messages to existing packages or create new packages
+with new versions of the messages. Also, we are not restricting the addition of cases to `oneof`s or values to `enum`s.
+All of these cases are easier to detect at runtime with standard `switch` statements than the addition of new fields.
+
+#### Additional Linting Rules
+
+The following additional rules will be enforced by a linter that
+complements [buf lint](https://docs.buf.build/lint/overview):
+
+* all message fields must be specified in continuous ascending order starting from `1`
+* all enums must be specified in continuous ascending order starting from `0` - otherwise it is too complex to check at
+  runtime whether an enum value is unknown. An alternative would be to make adding new values to existing enums breaking
+* all enum values must be `<= 255`. Any enum in a blockchain application which needs more than 256 values is probably
+  doing something very wrong.
+* all oneof's must be the *only* element in their containing message and must start at field number `1` and be added in
+  continuous ascending order - this makes it possible to quickly check for unknown values
+* all `oneof` field numbers must be `<= 255`. Any `oneof` which needs more field cases is probably doing something very
+  wrong.
+
+These requirements make the encoding and generated code simpler.
+
+### Encoding
+
+#### Buffers and Memory Management
+
+By default, this encoding attempts to use a single fixed size encoding buffer of 64kb. This imposes a limit on the
+maximum size of a message that can be encoded. In the context of a message passing protocol for blockchains, this
+is generally a reasonable limit and the only known valid use case for exceeding it is to store user-uploaded byte
+code for execution in VMs. To accommodate this, large `string` and `bytes` values can be encoded in additional
+standalone buffers if needed. Still, the body of a message included all scalar and message fields
+must fit inside the 64kb buffer.
```

Comment:
You simply can't assume that encoded messages can be represented in 64KB buffers.

Reply:
We can get statistics from existing cosmos zones to check how frequently larger buffers would be needed and figure out how to accomodate those cases